src/Polysemy/Internal/Union.hs

```diff
@@ -131,102 +138,130 @@ type Member e r = MemberNoError e r
--- | A singleton for 'Nat'.
-data SNat :: Nat -> Type where
-  SZ :: SNat 'Z
-  SS :: SNat n -> SNat ('S n)
-
-instance TestEquality SNat where
-  testEquality SZ     SZ     = Just Refl
-  testEquality (SS _) SZ     = Nothing
-  testEquality SZ     (SS _) = Nothing
-  testEquality (SS n) (SS m) =
-    case testEquality n m of
-      Nothing -> Nothing
-      Just Refl -> Just Refl
-  {-# INLINE testEquality #-}
-
-
-type family IndexOf (ts :: [k]) (n :: Nat) :: k where
-  IndexOf (k ': ks) 'Z = k
-  IndexOf (k ': ks) ('S n) = IndexOf ks n
-
-
-type family Found (ts :: [k]) (t :: k) :: Nat where
+-- | A proof that @e@ is an element of @r@.
+--
+-- Due to technical reasons, @'ElemOf' r e@ is not powerful enough to
+-- prove @'Member' e r@; however, it can still be used send actions of @e@
+-- into @r@ by using 'Polysemy.Internal.subsumeUsing'.
+data ElemOf r e where
+  Here :: ElemOf (e ': r) e
+  In   :: ElemOf r e -> ElemOf (e' ': r) e
+
+instance TestEquality (ElemOf r) where
+  testEquality Here Here
+    = Just Refl
+  testEquality (In e) (In e')
+    = testEquality e e'
+  testEquality _ _ =
+    Nothing
+
+type family LocateEffect (ts :: [k]) (t :: k) :: () where
 #ifndef NO_ERROR_MESSAGES
-  Found '[]       t = UnhandledEffect t
+  LocateEffect '[] t = UnhandledEffect t
 #endif
-  Found (t ': ts) t = 'Z
-  Found (u ': ts) t = 'S (Found ts t)
-
+  LocateEffect (t ': ts) t = '()
+  LocateEffect (u ': ts) t = LocateEffect ts t
 
 class Find (r :: [k]) (t :: k) where
-  finder :: SNat (Found r t)
+  membership' :: ElemOf r t
 
 instance {-# OVERLAPPING #-} Find (t ': z) t where
-  finder = SZ
-  {-# INLINE finder #-}
+  membership' = Here
+  {-# INLINE membership' #-}
+
+instance Find z t => Find (_1 ': z) t where
+  membership' = In $ membership' @_ @z @t
+  {-# INLINE membership' #-}
+
+------------------------------------------------------------------------------
+-- | A class for effect rows whose elements are inspectable.
+--
+-- This constraint is eventually satisfied as @r@ is instantied to a
+-- monomorphic list.
+-- (E.g when @r@ becomes something like
+-- @'[State Int, Output String, Embed IO]@)
+class KnownRow r where
+  tryMembership' :: forall e. Typeable e => Maybe (ElemOf r e)
+
+instance KnownRow '[] where
+  tryMembership' = Nothing
+  {-# INLINE tryMembership' #-}
+
+instance (Typeable e, KnownRow r) => KnownRow (e ': r) where
+  tryMembership' :: forall e'. Typeable e' => Maybe (ElemOf (e ': r) e')
+  tryMembership' = case eqT @e @e' of
+    Just Refl -> Just Here
+    _         -> In <$> tryMembership' @r @e'
+  {-# INLINE tryMembership' #-}
 
-instance ( Find z t
-         , Found (_1 ': z) t ~ 'S (Found z t)
-         ) => Find (_1 ': z) t where
-  finder = SS $ finder @_ @z @t
-  {-# INLINE finder #-}
+------------------------------------------------------------------------------
+-- | Given @'Member' e r@, extract a proof that @e@ is an element of @r@.
+membership :: Member e r => ElemOf r e
+membership = membership'
```

Comment:
Why the indirection?

Reply:
Partly to get right order on type variables, but mainly so that we have `Member` as the constraint, rather than `Finder`.